# StarRetroCopy

![Last commit](https://img.shields.io/github/last-commit/lindenb/jvarkit.png)

Scan retrocopies from the star-aligner/bwa output


## Usage

```
Usage: starretrocopy [options] Files
  Options:
    --bedpe, -P, -J
      Optional. Save possible sites of insertion in this Bed-PE file.
  * -gtf, --gtf
      A GTF (General Transfer Format) file. See 
      https://www.ensembl.org/info/website/upload/gff.html .
    -h, --help
      print help and exit
    --helpFormat
      What kind of help. One of [usage,markdown,xml].
    --mapq, -mapq
      Min mapping quality
      Default: 1
    -n, --min-cigar-size
      Minimal cigar element length for clipped sequence.
      Default: 6
    --min-depth, -D
      In a transcript one must found at least 'D' reads with a clip-length> 
      'min-cigar-size'. 
      Default: 1
    --no-ji, --no-star, --bwa
      Do not use the 'Ji' Attribute generated by STAR aligner to get the 
      intron bounds but use the CIGAR string (e.g: output of bwa)
      Default: false
    -o, --output
      Output file. Optional . Default: stdout
    -r, -R, --reference
      Indexed fasta Reference file. This file must be indexed with samtools 
      faidx and with picard CreateSequenceDictionary Can be used for reading 
      CRAM. 
    --bam, --save-bam
      Optional: save matching reads in this bam file. Can be used to visualize 
      the evidences in IGV.
    --transcript-id
      Limit to those transcript ID (one id per line)
    --version
      print version and exit
    --bai, -bai, --with-bai
      Use random access BAM using the bai and using the GTF data. May be slow 
      at startup
      Default: false

```


## Keywords

 * sam
 * bam
 * cigar
 * clip
 * sv
 * retrocopy
 * star


## Compilation

### Requirements / Dependencies

* java [compiler SDK 11](https://jdk.java.net/11/). Please check that this java is in the `${PATH}`. Setting JAVA_HOME is not enough : (e.g: https://github.com/lindenb/jvarkit/issues/23 )


### Download and Compile

```bash
$ git clone "https://github.com/lindenb/jvarkit.git"
$ cd jvarkit
$ ./gradlew starretrocopy
```

The java jar file will be installed in the `dist` directory.


## Creation Date

20190710

## Source code 

[https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/retrocopy/StarRetroCopy.java](https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/retrocopy/StarRetroCopy.java)

### Unit Tests

[https://github.com/lindenb/jvarkit/tree/master/src/test/java/com/github/lindenb/jvarkit/tools/retrocopy/StarRetroCopyTest.java](https://github.com/lindenb/jvarkit/tree/master/src/test/java/com/github/lindenb/jvarkit/tools/retrocopy/StarRetroCopyTest.java)


## Contribute

- Issue Tracker: [http://github.com/lindenb/jvarkit/issues](http://github.com/lindenb/jvarkit/issues)
- Source Code: [http://github.com/lindenb/jvarkit](http://github.com/lindenb/jvarkit)

## License

The project is licensed under the MIT license.

## Citing

Should you cite **starretrocopy** ? [https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md](https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md)

The current reference is:

[http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)

> Lindenbaum, Pierre (2015): JVarkit: java-based utilities for Bioinformatics. figshare.
> [http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)




