# VcfEnsemblVepRest

![Last commit](https://img.shields.io/github/last-commit/lindenb/jvarkit.png)

Annotate a VCF with ensembl REST API


## Usage

```
Usage: vcfensemblvep [options] Files
  Options:
    -n, --batchSize
      batch size. How many variant to send in one HTTP query
      Default: 100
    -e, --extension
      REST server URL Path extension
      Default: /vep/homo_sapiens/region
    -h, --help
      print help and exit
    --helpFormat
      What kind of help. One of [usage,markdown,xml].
    -o, --output
      Output file. Optional . Default: stdout
    -s, --server
      Ensemble REST server url.
      Default: http://grch37.rest.ensembl.org
    -t, --tags, --tag
      [20180214] comma/space separated of tag that will be generated by the 
      XSL stylesheet.
      Default: VEPREST_TRANSCRIPT,VEPREST_COLOCATED_VARIANTS,VEPREST_INTERGENIC_CSQ,VEPREST_REGULATORY_CSQ
    -T, --tee
      'Tee' xml response to stderr
      Default: false
    -x, --xslt, --template
      [20180214] XSLT stylesheet template that will be used to transform the 
      XML response from ensembl to a set of lines.Each line in the generated 
      document must starts with a tag (TAG=...) defined with the --tag 
      section. Empty lines or lines starting with a hash will be ignored. The 
      XSL output method must be text. The processed node is a <data> node in 
      the Rest response. Example at: 
      https://gist.github.com/lindenb/12dea8e701d18280e8cb0b65270064f6 .The 
      default XSL template is embedded in the code: https://github.com/lindenb/jvarkit/blob/master/src/main/java/com/github/lindenb/jvarkit/tools/ensembl/VcfEnsemblVepRest.java
    --version
      print version and exit
    -nofail
      [20180213] Do not fail on network error
      Default: false

```


## Keywords

 * vcf
 * annotation
 * rest
 * ensembl
 * xml
 * xslt
 * xsl


## Compilation

### Requirements / Dependencies

* java [compiler SDK 11](https://jdk.java.net/11/). Please check that this java is in the `${PATH}`. Setting JAVA_HOME is not enough : (e.g: https://github.com/lindenb/jvarkit/issues/23 )


### Download and Compile

```bash
$ git clone "https://github.com/lindenb/jvarkit.git"
$ cd jvarkit
$ ./gradlew vcfensemblvep
```

The java jar file will be installed in the `dist` directory.

## Source code 

[https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/ensembl/VcfEnsemblVepRest.java](https://github.com/lindenb/jvarkit/tree/master/src/main/java/com/github/lindenb/jvarkit/tools/ensembl/VcfEnsemblVepRest.java)

### Unit Tests

[https://github.com/lindenb/jvarkit/tree/master/src/test/java/com/github/lindenb/jvarkit/tools/ensembl/VcfEnsemblVepRestTest.java](https://github.com/lindenb/jvarkit/tree/master/src/test/java/com/github/lindenb/jvarkit/tools/ensembl/VcfEnsemblVepRestTest.java)


## Contribute

- Issue Tracker: [http://github.com/lindenb/jvarkit/issues](http://github.com/lindenb/jvarkit/issues)
- Source Code: [http://github.com/lindenb/jvarkit](http://github.com/lindenb/jvarkit)

## License

The project is licensed under the MIT license.

## Citing

Should you cite **vcfensemblvep** ? [https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md](https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md)

The current reference is:

[http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)

> Lindenbaum, Pierre (2015): JVarkit: java-based utilities for Bioinformatics. figshare.
> [http://dx.doi.org/10.6084/m9.figshare.1425030](http://dx.doi.org/10.6084/m9.figshare.1425030)




## Example

```bash

 $ wget -q -O - "http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20110521/ALL.chr1.phase1_release_v3.20101123.snps_indels_svs.genotypes.vcf.gz"  |\
  	gunzip -c  |\
  	java -jar dist/vcfensemblvep.jar  -n 10 | grep -v '^#' | cut -f 1,2,4,5,8 | head

1	10583	G	A	AA=.;AC=314;AF=0.14;AFR_AF=0.04;AMR_AF=0.17;AN=2184;ASN_AF=0.13;AVGPOST=0.7707;ERATE=0.0161;EUR_AF=0.21;LDAF=0.2327;RSQ=0.4319;SNPSOURCE=LOWCOV;THETA=0.0046;VEPREST_COLOCATED_VARIANTS=allele_string|G/A|end|10583|id|rs58108140|seq_region_name|1|start|10583|strand|1;VEPREST_TRANSCRIPT=biotype|unprocessed_pseudogene|distance|3780|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000423562|variant_allele|A|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|canonical|1|distance|3780|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000438504|variant_allele|A|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|distance|1427|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|impact|MODIFIER|strand|1|transcript_id|ENST00000450305|variant_allele|A|consequences|upstream_gene_variant,biotype|processed_transcript|canonical|1|distance|1286|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|impact|MODIFIER|strand|1|transcript_id|ENST00000456328|variant_allele|A|consequences|upstream_gene_variant,biotype|unprocessed_pseudogene|distance|3821|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000488147|variant_allele|A|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|distance|1289|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|impact|MODIFIER|strand|1|transcript_id|ENST00000515242|variant_allele|A|consequences|upstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|distance|1291|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|impact|MODIFIER|strand|1|transcript_id|ENST00000518655|variant_allele|A|consequences|upstream_gene_variant,biotype|unprocessed_pseudogene|distance|3828|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000538476|variant_allele|A|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|distance|3780|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000541675|variant_allele|A|consequences|downstream_gene_variant;VT=SNP
1	10611	C	G	AA=.;AC=41;AF=0.02;AFR_AF=0.01;AMR_AF=0.03;AN=2184;ASN_AF=0.01;AVGPOST=0.9330;ERATE=0.0048;EUR_AF=0.02;LDAF=0.0479;RSQ=0.3475;SNPSOURCE=LOWCOV;THETA=0.0077;VEPREST_COLOCATED_VARIANTS=allele_string|C/G|end|10611|id|rs189107123|seq_region_name|1|start|10611|strand|1;VEPREST_TRANSCRIPT=biotype|unprocessed_pseudogene|distance|3752|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000423562|variant_allele|G|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|canonical|1|distance|3752|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000438504|variant_allele|G|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|distance|1399|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|impact|MODIFIER|strand|1|transcript_id|ENST00000450305|variant_allele|G|consequences|upstream_gene_variant,biotype|processed_transcript|canonical|1|distance|1258|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|impact|MODIFIER|strand|1|transcript_id|ENST00000456328|variant_allele|G|consequences|upstream_gene_variant,biotype|unprocessed_pseudogene|distance|3793|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000488147|variant_allele|G|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|distance|1261|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|impact|MODIFIER|strand|1|transcript_id|ENST00000515242|variant_allele|G|consequences|upstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|distance|1263|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|impact|MODIFIER|strand|1|transcript_id|ENST00000518655|variant_allele|G|consequences|upstream_gene_variant,biotype|unprocessed_pseudogene|distance|3800|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000538476|variant_allele|G|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|distance|3752|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000541675|variant_allele|G|consequences|downstream_gene_variant;VT=SNP
1	13302	C	T	AA=.;AC=249;AF=0.11;AFR_AF=0.21;AMR_AF=0.08;AN=2184;ASN_AF=0.02;AVGPOST=0.8895;ERATE=0.0058;EUR_AF=0.14;LDAF=0.1573;RSQ=0.6281;SNPSOURCE=LOWCOV;THETA=0.0048;VEPREST_COLOCATED_VARIANTS=allele_string|C/G/T|end|13302|id|rs75241669|seq_region_name|1|start|13302|strand|1;VEPREST_TRANSCRIPT=biotype|unprocessed_pseudogene|distance|1061|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000423562|variant_allele|T|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|canonical|1|distance|1061|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000438504|variant_allele|T|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|cdna_end|342|cdna_start|342|exon|5/6|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000450305.2:n.342C>T|impact|MODIFIER|strand|1|transcript_id|ENST00000450305|variant_allele|T|consequences|non_coding_transcript_exon_variant,biotype|processed_transcript|canonical|1|cdna_end|550|cdna_start|550|exon|3/3|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000456328.2:n.550C>T|impact|MODIFIER|strand|1|transcript_id|ENST00000456328|variant_allele|T|consequences|non_coding_transcript_exon_variant,biotype|unprocessed_pseudogene|distance|1102|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000488147|variant_allele|T|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|cdna_end|543|cdna_start|543|exon|3/3|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000515242.2:n.543C>T|impact|MODIFIER|strand|1|transcript_id|ENST00000515242|variant_allele|T|consequences|non_coding_transcript_exon_variant,biotype|transcribed_unprocessed_pseudogene|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000518655.2:n.482-101C>T|impact|MODIFIER|intron|2/3|strand|1|transcript_id|ENST00000518655|variant_allele|T|consequences|intron_variant&non_coding_transcript_variant,biotype|unprocessed_pseudogene|distance|1109|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000538476|variant_allele|T|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|distance|1061|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000541675|variant_allele|T|consequences|downstream_gene_variant;VT=SNP
1	13327	G	C	AA=.;AC=59;AF=0.03;AFR_AF=0.02;AMR_AF=0.03;AN=2184;ASN_AF=0.02;AVGPOST=0.9698;ERATE=0.0012;EUR_AF=0.04;LDAF=0.0359;RSQ=0.6482;SNPSOURCE=LOWCOV;THETA=0.0204;VEPREST_TRANSCRIPT=biotype|unprocessed_pseudogene|distance|1036|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000423562|variant_allele|C|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|canonical|1|distance|1036|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000438504|variant_allele|C|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|cdna_end|367|cdna_start|367|exon|5/6|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000450305.2:n.367G>C|impact|MODIFIER|strand|1|transcript_id|ENST00000450305|variant_allele|C|consequences|non_coding_transcript_exon_variant,biotype|processed_transcript|canonical|1|cdna_end|575|cdna_start|575|exon|3/3|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000456328.2:n.575G>C|impact|MODIFIER|strand|1|transcript_id|ENST00000456328|variant_allele|C|consequences|non_coding_transcript_exon_variant,biotype|unprocessed_pseudogene|distance|1077|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000488147|variant_allele|C|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|cdna_end|568|cdna_start|568|exon|3/3|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000515242.2:n.568G>C|impact|MODIFIER|strand|1|transcript_id|ENST00000515242|variant_allele|C|consequences|non_coding_transcript_exon_variant,biotype|transcribed_unprocessed_pseudogene|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000518655.2:n.482-76G>C|impact|MODIFIER|intron|2/3|strand|1|transcript_id|ENST00000518655|variant_allele|C|consequences|intron_variant&non_coding_transcript_variant,biotype|unprocessed_pseudogene|distance|1084|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000538476|variant_allele|C|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|distance|1036|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000541675|variant_allele|C|consequences|downstream_gene_variant;VT=SNP
1	13957	TC	T	AA=TC;AC=35;AF=0.02;AFR_AF=0.02;AMR_AF=0.02;AN=2184;ASN_AF=0.01;AVGPOST=0.8711;ERATE=0.0065;EUR_AF=0.02;LDAF=0.0788;RSQ=0.2501;THETA=0.0100;VEPREST_COLOCATED_VARIANTS=allele_string|C/-|end|13958|id|rs201747181|seq_region_name|1|start|13958|strand|1;VEPREST_TRANSCRIPT=biotype|unprocessed_pseudogene|distance|405|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000423562|variant_allele|-|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|canonical|1|distance|405|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000438504|variant_allele|-|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|distance|288|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|impact|MODIFIER|strand|1|transcript_id|ENST00000450305|variant_allele|-|consequences|downstream_gene_variant,biotype|processed_transcript|canonical|1|cdna_end|1206|cdna_start|1206|exon|3/3|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvs_offset|4|hgvsc|ENST00000456328.2:n.1210del|impact|MODIFIER|strand|1|transcript_id|ENST00000456328|variant_allele|-|consequences|non_coding_transcript_exon_variant,biotype|unprocessed_pseudogene|distance|446|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000488147|variant_allele|-|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|cdna_end|1199|cdna_start|1199|exon|3/3|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvs_offset|4|hgvsc|ENST00000515242.2:n.1203del|impact|MODIFIER|strand|1|transcript_id|ENST00000515242|variant_allele|-|consequences|non_coding_transcript_exon_variant,biotype|transcribed_unprocessed_pseudogene|cdna_end|1032|cdna_start|1032|exon|4/4|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvs_offset|4|hgvsc|ENST00000518655.2:n.1036del|impact|MODIFIER|strand|1|transcript_id|ENST00000518655|variant_allele|-|consequences|non_coding_transcript_exon_variant,biotype|unprocessed_pseudogene|distance|453|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000538476|variant_allele|-|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|distance|405|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000541675|variant_allele|-|consequences|downstream_gene_variant;VT=INDEL
1	13980	T	C	AA=.;AC=45;AF=0.02;AFR_AF=0.01;AMR_AF=0.02;AN=2184;ASN_AF=0.02;AVGPOST=0.9221;ERATE=0.0034;EUR_AF=0.02;LDAF=0.0525;RSQ=0.3603;SNPSOURCE=LOWCOV;THETA=0.0139;VEPREST_TRANSCRIPT=biotype|unprocessed_pseudogene|distance|383|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000423562|variant_allele|C|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|canonical|1|distance|383|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000438504|variant_allele|C|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|distance|310|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|impact|MODIFIER|strand|1|transcript_id|ENST00000450305|variant_allele|C|consequences|downstream_gene_variant,biotype|processed_transcript|canonical|1|cdna_end|1228|cdna_start|1228|exon|3/3|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000456328.2:n.1228T>C|impact|MODIFIER|strand|1|transcript_id|ENST00000456328|variant_allele|C|consequences|non_coding_transcript_exon_variant,biotype|unprocessed_pseudogene|distance|424|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000488147|variant_allele|C|consequences|downstream_gene_variant,biotype|transcribed_unprocessed_pseudogene|cdna_end|1221|cdna_start|1221|exon|3/3|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000515242.2:n.1221T>C|impact|MODIFIER|strand|1|transcript_id|ENST00000515242|variant_allele|C|consequences|non_coding_transcript_exon_variant,biotype|transcribed_unprocessed_pseudogene|cdna_end|1054|cdna_start|1054|exon|4/4|gene_id|ENSG00000223972|gene_symbol|DDX11L1|gene_symbol_source|HGNC|hgnc_id|37102|hgvsc|ENST00000518655.2:n.1054T>C|impact|MODIFIER|strand|1|transcript_id|ENST00000518655|variant_allele|C|consequences|non_coding_transcript_exon_variant,biotype|unprocessed_pseudogene|distance|431|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000538476|variant_allele|C|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|distance|383|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000541675|variant_allele|C|consequences|downstream_gene_variant;VT=SNP
1	30923	G	T	AA=T;AC=1584;AF=0.73;AFR_AF=0.48;AMR_AF=0.80;AN=2184;ASN_AF=0.89;AVGPOST=0.7335;ERATE=0.0183;EUR_AF=0.73;LDAF=0.6576;RSQ=0.5481;SNPSOURCE=LOWCOV;THETA=0.0162;VEPREST_COLOCATED_VARIANTS=afr_allele|T|afr_maf|0.6687|allele_string|G/T|amr_allele|T|amr_maf|0.9164|eas_allele|T|eas_maf|0.996|end|30923|eur_allele|T|eur_maf|0.9364|id|rs806731|minor_allele|G|minor_allele_freq|0.1276|sas_allele|T|sas_maf|0.9233|seq_region_name|1|start|30923|strand|1;VEPREST_TRANSCRIPT=biotype|lincRNA|canonical|1|distance|3631|gene_id|ENSG00000237613|gene_symbol|FAM138A|gene_symbol_source|HGNC|hgnc_id|32334|impact|MODIFIER|strand|-1|transcript_id|ENST00000417324|variant_allele|T|consequences|downstream_gene_variant,biotype|unprocessed_pseudogene|distance|1553|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000423562|variant_allele|T|consequences|upstream_gene_variant,biotype|unprocessed_pseudogene|canonical|1|distance|1553|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000438504|variant_allele|T|consequences|upstream_gene_variant,biotype|lincRNA|distance|4322|gene_id|ENSG00000237613|gene_symbol|FAM138A|gene_symbol_source|HGNC|hgnc_id|32334|impact|MODIFIER|strand|-1|transcript_id|ENST00000461467|variant_allele|T|consequences|downstream_gene_variant,biotype|lincRNA|gene_id|ENSG00000243485|gene_symbol|MIR1302-10|gene_symbol_source|HGNC|hgnc_id|38233|hgvsc|ENST00000469289.1:n.402-53G>T|impact|MODIFIER|intron|1/1|strand|1|transcript_id|ENST00000469289|variant_allele|T|consequences|intron_variant&non_coding_transcript_variant,biotype|lincRNA|canonical|1|gene_id|ENSG00000243485|gene_symbol|MIR1302-10|gene_symbol_source|HGNC|hgnc_id|38233|hgvsc|ENST00000473358.1:n.591-53G>T|impact|MODIFIER|intron|2/2|strand|1|transcript_id|ENST00000473358|variant_allele|T|consequences|intron_variant&non_coding_transcript_variant,biotype|unprocessed_pseudogene|distance|1353|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000488147|variant_allele|T|consequences|upstream_gene_variant,biotype|unprocessed_pseudogene|distance|1117|gene_id|ENSG00000227232|gene_symbol|WASH7P|gene_symbol_source|HGNC|hgnc_id|38034|impact|MODIFIER|strand|-1|transcript_id|ENST00000538476|variant_allele|T|consequences|upstream_gene_variant,biotype|miRNA|distance|420|gene_id|ENSG00000243485|gene_symbol|MIR1302-10|gene_symbol_source|HGNC|hgnc_id|38233|impact|MODIFIER|strand|1|transcript_id|ENST00000607096|variant_allele|T|consequences|downstream_gene_variant;VT=SNP
1	46402	C	CTGT	AA=.;AC=8;AF=0.0037;AFR_AF=0.01;AN=2184;ASN_AF=0.0017;AVGPOST=0.8325;ERATE=0.0072;LDAF=0.0903;RSQ=0.0960;THETA=0.0121;VEPREST_COLOCATED_VARIANTS=allele_string|-/TGT|end|46402|id|rs199681827|seq_region_name|1|start|46403|strand|1;VEPREST_INTERGENIC_CSQ=impact|MODIFIER|variant_allele|TGT|consequences|intergenic_variant;VT=INDEL
1	47190	G	GA	AA=G;AC=29;AF=0.01;AFR_AF=0.06;AMR_AF=0.0028;AN=2184;AVGPOST=0.9041;ERATE=0.0041;LDAF=0.0628;RSQ=0.2883;THETA=0.0153;VEPREST_COLOCATED_VARIANTS=allele_string|-/A|end|47190|id|rs200430748|seq_region_name|1|start|47191|strand|1;VEPREST_INTERGENIC_CSQ=impact|MODIFIER|variant_allele|A|consequences|intergenic_variant;VT=INDEL
1	51476	T	C	AA=C;AC=18;AF=0.01;AFR_AF=0.01;AMR_AF=0.01;AN=2184;ASN_AF=0.01;AVGPOST=0.9819;ERATE=0.0021;EUR_AF=0.01;LDAF=0.0157;RSQ=0.5258;SNPSOURCE=LOWCOV;THETA=0.0103;VEPREST_TRANSCRIPT=biotype|unprocessed_pseudogene|distance|1573|gene_id|ENSG00000268020|gene_symbol|OR4G4P|gene_symbol_source|HGNC|hgnc_id|14822|impact|MODIFIER|strand|1|transcript_id|ENST00000594647|variant_allele|C|consequences|upstream_gene_variant,biotype|unprocessed_pseudogene|canonical|1|distance|997|gene_id|ENSG00000268020|gene_symbol|OR4G4P|gene_symbol_source|HGNC|hgnc_id|14822|impact|MODIFIER|strand|1|transcript_id|ENST00000606857|variant_allele|C|consequences|upstream_gene_variant;VT=SNP

```

## History

* 2018-02-13: removed XSD, parsing DOM. Added SNPEFF output (but it's incomplete for now...)

